lib/room_booker.rb

```diff
@@ -0,0 +1,172 @@
+require 'date'
+require_relative 'date_logic'
+require_relative 'reservation'
+require_relative 'room_block'
+require_relative 'room_not_available'
+
+require 'pry'
+
+module BookingLogic
+  class RoomBooker
+    Room = Struct.new(:id, :cost, :block_reserved)
+
+    attr_reader :rooms
+    attr_accessor :reservations, :blocks
+
+    def initialize
+      @rooms = populate_rooms
+      @reservations = []
+      @blocks = []
+    end
+
+    def populate_rooms
+      rooms_array = []
+
+      20.times do |id|
+        room = Room.new((id + 1), 200)
+        rooms_array << room
+      end
+
+      return rooms_array
+    end
+
+    def list_rooms
+      return rooms
+    end
+
+    def find_room_by_id(room_id)
+      return rooms.find { |room| room.id == room_id }
+    end
+
+    def list_reservations(date)
+      list_of_reservations = []
+
+      reservations.each do |reservation|
+        if DateLogic.date_range_include?(reservation, date)
+          list_of_reservations << reservation
+        end
+      end
+
+      return list_of_reservations
+    end
+
+    def list_available_rooms(check_in, check_out)
+      reserved_rooms = []
+
+      reservations.each do |reservation|
+        unless DateLogic.date_ranges_exclusive?(
+          reservation.check_in,
+          reservation.check_out,
+          check_in,
+          check_out
+        )
+
+          reserved_rooms << reservation.room
+        end
+      end
+
+      unavailable_rooms = reserved_rooms + find_blocked_rooms(check_in, check_out)
+
+      available_rooms = rooms.dup
+
+      unavailable_rooms.each do |unavailable_room|
+        available_rooms.delete_if { |room| room.id == unavailable_room.id }
+      end
+
+      return available_rooms
+    end
+
+    def new_reservation(room_id, check_in, check_out)
+      room = find_room_by_id(room_id)
+      room_unavailable?(room, check_in, check_out)
+      new_reservation = BookingLogic::Reservation.new(room, check_in, check_out)
+      reservations << new_reservation
+      return new_reservation
```

Comment:
I just want to call out that we never formally teach this pattern, but this is a great and common pattern: when a method makes something new, it returns that new thing in the end. Nice!